### PR TITLE
LibVNCClient: limit security type count to 254

### PR DESCRIPTION
**Pull request by tobydox**
As the security type count is stored in an uint8_t it will never exceed
255. The check for too many security types will therefore always
evaluate to FALSE. We can prevent this by limiting the number of
accepted security types to 254.

### CODE REVIEW

**Comment by tobydox** [comment]
Patch updated to work with MSVC + elimination of extra constant. Thanks in advance for reviewing!

File: libvncclient/tls_gnutls.c [hunk 1]
```diff
@@ -316,7 +316,7 @@ ReadVeNCryptSecurityType(rfbClient* client, uint32_t *result)
     uint8_t count=0;
     uint8_t loop=0;
     uint8_t flag=0;
-    uint32_t tAuth[256], t;
+    uint32_t tAuth[254], t;
```
Comment (rgacogne) [inline libvncclient/tls_gnutls.c]: While I agree that there will very likely never be more than 254 types advertised, setting an arbitrary limit feels a bit wrong. Perhaps we could cast `count` to `size_t` in the comparison instead?

File: libvncclient/tls_gnutls.c [hunk 2]
```diff
@@ -328,7 +328,7 @@ ReadVeNCryptSecurityType(rfbClient* client, uint32_t *result)
         return FALSE;
     }
 
-    if (count>sizeof(tAuth))
+    if (count>(sizeof(tAuth)/sizeof(tAuth[0])))
     {
         rfbClientLog("%d security types are too many; maximum is %d\n", count, sizeof(tAuth));
```
Comment (rgacogne) [inline libvncclient/tls_gnutls.c]: We need to fix the maximum size in the message as well:

```suggestion
         rfbClientLog("%d security types are too many; maximum is %d\n", count, (sizeof(tAuth)/sizeof(tAuth[0])));
```

File: libvncclient/tls_openssl.c [hunk 3]
```diff
@@ -454,7 +454,7 @@ ReadVeNCryptSecurityType(rfbClient* client, uint32_t *result)
         return FALSE;
     }
 
-    if (count>sizeof(tAuth))
+    if (count>(sizeof(tAuth)/sizeof(tAuth[0])))
     {
         rfbClientLog("%d security types are too many; maximum is %d\n", count, sizeof(tAuth));
```
Comment (rgacogne) [inline libvncclient/tls_openssl.c]:
```suggestion
         rfbClientLog("%d security types are too many; maximum is %d\n", count, (sizeof(tAuth)/sizeof(tAuth[0])));
```

File: libvncclient/tls_gnutls.c (hunk 1 above)
Comment (tobydox) [reply libvncclient/tls_gnutls.c]: It wouldn't make any difference ;-) Per RFB protocol the number of security types is transmitted as a uint8_t so even casting wouldn't change the behaviour. Either we drop the check completely (or check against 0 only) and keep `tAuth` at 256 elements or we choose an arbitrary limit as proposed. Any opinions?

Comment (rgacogne) [reply libvncclient/tls_gnutls.c]: It would make the warning go away :)

**Pull request closed without merging**